File: backsplash.py

class Tile(object):
    
    def __init__(self, height, width, row=None, col=None, **kw):
        self.width = width
        self.height = height
        self.row = row
        self.col = col
        self.__dict__.update(kw)

    def copy(self):
        return self.__class__(**self.__dict__)


class Mosaic(object):

    def __init__(self, width, tile_size=100, padding=0):
        self._width = width
        self._holes = []
        self._tiles = {}
        self._tile_size = tile_size
        self._padding = padding

    @property
    def width(self):
        return self._width

    @property
    def height(self):
        return len(self._holes)

    def css(self, tile):
        return {
            'top': tile.row * (self._tile_size + self._padding),
            'left': tile.col * (self._tile_size + self._padding),
            'height': tile.height * (self._tile_size + self._padding) - self._padding,
            'width': tile.width * (self._tile_size + self._padding) - self._padding,
        }

    def add(self, tile=None, **kw):
        tile = tile.copy() or Tile(**kw)
        if tile.row is None or tile.col is None:
            self._find_hole(tile)
        self._occupy(tile)
        return tile

    def linearize(self):
        seen = set()
        for r in xrange(len(self._holes)):
            for c in xrange(self._width):
                t = self._tiles.get((r, c))
                if not t:
                    continue
                if t not in seen:
                    yield t
                    seen.add(t)

    def _find_hole(self, tile):
        for r in xrange(len(self._holes)):
            for c in xrange(self._width):
                if self._holes[r] and self._holes[r][c]:
                    if self._fits_holes(tile, r, c):
                        tile.row = r
                        tile.col = c
                        return tile
        tile.row = len(self._holes)
        tile.col = 0
        return tile

    def _fits_holes(self, tile, r, c):
        for R in xrange(r, r + tile.height):
            for C in xrange(c, c + tile.width):
                if C >= self._width:
                    return False
                if R < len(self._holes) and (not self._holes[R] or not self._holes[R][C]):
                    return False
        return True

    def _occupy(self, tile):
        while len(self._holes) < tile.row + tile.height:
            self._holes.append([True] * self._width)
        for r in xrange(tile.row, tile.row + tile.height):
            for c in xrange(tile.col, tile.col + tile.width):
                self._holes[r][c] = False
                self._tiles[(r,c)] = tile
            if not any(self._holes[r]):
                self._holes[r] = False

File: backsplash.js


var backsplash = (function(){

function clone(obj) {

    if(obj == null || typeof(obj) != 'object')
        return obj;

    var temp = {}; // obj.constructor();

    for(var key in obj) {
        if(obj.hasOwnProperty(key)) {
            temp[key] = obj[key];
        }
    }
    return temp;

}


var Layout = function(opts) {

    this.opts = clone(opts || {});

    this.width = this.opts.width || 5;
    this.tile_size = this.opts.tile_size || 100;
    this.padding = this.opts.padding || 0;

    this.tiles = [];
    this.holes = [];

}


Layout.prototype.css = function(tile) {
    return {
        'top': tile.row * (this.tile_size + this.padding),
        'left': tile.col * (this.tile_size + this.padding),
        'height': tile.height * (this.tile_size + this.padding) - this.padding,
        'width': tile.width * (this.tile_size + this.padding) - this.padding,
    }
}


Layout.prototype.linearize = function() {
    var res = []
    var seen = {}
    for (var row = 0; row < this.tiles.length; row++) {
        for (var col = 0; col < this.width; col++) {
            var tile = this.tiles[row][col];
            if (!tile) {
                continue;
            }
            var key = this.width * tile.row + tile.col;
            if (!seen[key]) {
                res.push(tile);
                seen[key] = true;
            }
        }
    }
    return res;
}


Layout.prototype.add = function(tile, opts) {

    tile = clone(tile);
    opts = opts ? opts : {};

    if (opts.ignore_position) {
        tile.row = tile.col = null;
    }

    if (!(typeof(tile.row) === "number" && typeof(tile.col) === "number")) {
        this._find_hole(tile);
    }
    this.occupy(tile);
    return tile;

}


Layout.prototype._find_hole = function(tile) {

    for (var r = 0; r < this.holes.length; r++) {
        for (var c = 0; c < this.width; c++) {
            if (this.holes[r] && this.holes[r][c]) {
                if (this._fits_holes(tile, r, c)) {
                    tile.row = r;
                    tile.col = c;
                    return tile;
                }
            }
        }
    }

    tile.row = this.holes.length;
    tile.col = 0;
    return tile;

}


Layout.prototype._fits_holes = function(tile, r, c) {
    for (var R = r; R < r + tile.height; R++) {
        for (var C = c; C < c + tile.width; C++) {
            if (R < this.holes.length && (!this.holes[R] || !this.holes[R][C])) {
                return false;
            }
        }
    }
    return true;
}


Layout.prototype.occupy = function(tile) {

    while (this.holes.length < tile.row + tile.height + 2) {
        var holes = [];
        var tiles = [];
        for (var i = 0; i < this.width; i++) {
            holes.push(true);
            tiles.push(null);
        }
        this.holes.push(holes);
        this.tiles.push(tiles)
    }

    for (var r = tile.row; r < tile.row + tile.height; r++) {
        
        if (!this.holes[r] || !this.tiles[r]) {
            throw "invalid row " + r;
        }

        for (var c = tile.col; c < tile.col + tile.width; c++) {
            this.holes[r][c] = false;
            this.tiles[r][c] = tile;
        }

        // Here is the only real different between them.
        var empty = true;
        for (var c = 0; c < this.width; c++) {
            if (this.holes[r][c]) {
                empty = false;
                break;
            }
        }
        if (empty) {
            this.holes[r] = null;
        }

    }



}


return {
    _clone: clone,
    Layout: Layout
}

})();


if (typeof window === 'undefined') {
    module.exports = backsplash
}

